src/zcash/address/zip32.cpp

```diff
@@ -153,13 +297,73 @@ SaplingExtendedFullViewingKey SaplingExtendedSpendingKey::ToXFVK() const
     return ret;
 }
 
-libzcash::SaplingPaymentAddress SaplingExtendedSpendingKey::DefaultAddress() const
-{
-    return ToXFVK().DefaultAddress();
+//
+// Unified
+//
+
+std::optional<std::pair<UnifiedSpendingKey, HDKeyPath>> UnifiedSpendingKey::ForAccount(const HDSeed& seed, uint32_t bip44CoinType, uint32_t accountId) {
+    UnifiedSpendingKey usk;
+    usk.accountId = accountId;
+
+    auto transparentKey = DeriveZip32TransparentAccountKey(seed, bip44CoinType, accountId);
+    if (!transparentKey.has_value()) return std::nullopt;
+    usk.transparentKey = transparentKey.value().first;
+
+    auto saplingKey = SaplingExtendedSpendingKey::ForAccount(seed, bip44CoinType, accountId);
+    usk.saplingKey = saplingKey.first;
+
+    return std::make_pair(usk, saplingKey.second);
+}
+
+UnifiedFullViewingKey UnifiedSpendingKey::ToFullViewingKey() const {
+    UnifiedFullViewingKey ufvk;
+
+    if (transparentKey.has_value()) {
+        ufvk.transparentKey = transparentKey.value().Neuter();
+    }
+
+    if (saplingKey.has_value()) {
+        ufvk.saplingKey = saplingKey.value().ToXFVK();
+    }
+
+    return ufvk;
+}
+
+std::optional<ZcashdUnifiedAddress> UnifiedFullViewingKey::Address(diversifier_index_t j) const {
+    ZcashdUnifiedAddress ua;
+
+    if (transparentKey.has_value()) {
+        // ensure that the diversifier index is small enough for a t-addr
+        if (MAX_TRANSPARENT_CHILD_IDX < j) return std::nullopt;
+
+        CExtPubKey changeKey;
+        if (!transparentKey.value().Derive(changeKey, 0)) {
+            return std::nullopt;
+        }
+
+        CExtPubKey childKey;
+        unsigned int childIndex = (unsigned int) j.GetUint64(0);
```

Comment:
I'd personally prefer a method `std::optional<unsigned int> diversifier_index_t::ToTransparentChildIndex()` that performs this conversion (and the range check above). The `base_blob::GetUint64` function in particular is an implementation detail (the magic `0` in particular required me to go and look at how that function operated).

Reply:
Fixed in 867550d53ef2409b3cfbe8ea36310b570dc8988f